src/vs/vscode.d.ts

```diff
@@ -968,6 +968,17 @@ declare module 'vscode' {
 		edit(callback: (editBuilder: TextEditorEdit) => void, options?: { undoStopBefore: boolean; undoStopAfter: boolean; }): Thenable<boolean>;
 
 		/**
+		 * Inserts the given snippet template and enters snippet mode.
+		 *
+		 * If the editor is already in snippet mode, insertion fails and the returned promise resolves to false.
+		 *
+		 * @param template The snippet template to insert
+		 * @param posOrRange The position or replacement range representing the location of the insertion.
+		 * @return A promise that resolves with a value indicating if the snippet could be inserted.
+		 */
+		insertSnippet(template: string, posOrRange: Position | Range): Thenable<boolean>;
```

Comment:
> How should this work for multiple selections?
As with a single selection, the editor API allows to set one or many selections. The snippet logic tho isn't yet smart enough to always honour multiple cursors when snippets are inserted. We have a issue for this and it will be implemented in the near future

> how does entering snippet mode fit in with subsequent insertions, replacements, etc.? It seems like this might have more in common with CompletionItem than editing.

Sure, I was just wondering if reusing the name `edit` makes sense because that is sort of what we do. It doesn't have to fit in the builder but could just reuse the name, like so: 

```ts
edit(snippet: SnippetString, options?:{/*undo stops control*/}): void;

edit(callback: (editBuilder: TextEditorEdit) => void, options?: { undoStopBefore: boolean; undoStopAfter: boolean; }): Thenable<boolean>;
```

> Another thought: Should we supply a callback to know when snippet mode was exited?

Not a fan because it really depends on the user tabbing through things to exit snippet mode.